### ISSUE TITLE
Organizations / groups from LDAP

### ISSUE THREAD

**Issue by lbeltrame** [gogits/gogs#662]
As the subject says, this would be a nice feature to have. Rationale: as I said, KDE is evaluating Gogs for its public Git infrastructure, and this is one of the features that are important. Since th number of accounts are large (1000+) the idea would be to use LDAP to get groups ("Organizations" in gogs speak) and assign people to them.

For reference, this is a feature that GitLab has, but only it its for-pay variant.

**Comment by Unknwon** [comment]
What is the status of this issue?

**Comment by Unknwon** [comment]
Any updates?

**Comment by Unknwon** [comment]
Close due to lack of feedback.

**Comment by Fnux** [comment]
I would like to re-open this issue. It would be nice to be able to sync LDAP groups with gogs organizations. I'm thinking about replacing a gitlab instance by gogs and this feature would be very useful.

**Comment by Dasoren** [comment]
@Fnux If I might ask, why would you want to replace gitlab with gogs? Not to dis gogs, love it all day long. Just wondering what your thinking is here.

**Comment by Unknwon** [comment]
Thanks @Fnux !

Any idea of how to implement this?

**Comment by Fnux** [comment]
We should allow - on a specific LDAP source - to sync some groups with the local "Organizations". 
A way would be to add this part in each LDAP source configuration page. You should be able to add as many "links" as you can : just two fields per "link".

* the LDAP group
* the local "Organization"

And when an user sign up using this source, gogs should check if it belongs to the specified group. If that's the case, add him to the related organization. We also have to check if the user was removed from the specified group : check once every X hours ? First login of the day ? Every login ?

Sadly, I do not have the time/knowledge to do it by myself right now. (Maybe in late January/February?)

**Comment by Fnux** [comment]
@Dasoren Well, I used GitLab a while for myself and I'm actually maintaining a GitLab instance for an association. We're in an university. We have our own LDAP, the school has its LDAP and an authentification system called "Tequila". GitLab CE allows a single LDAP server, so I had to change few things in order to allow authentification through "Tequila". It's awful to maintain (really). It takes a lot of ressources. It's way to complex.

About *Gogs* : 
* Easy to maintain (:heart: that's the big point). No dependencies except database. 
* Way more simple than the whole GitLab-sutff.
* The interface is great.
* Use something like 50 times less resources

**Comment by Unknwon** [comment]
@Fnux great points!

Is it a one-way sync or two-way sync?

**Comment by nestabur** [comment]
+1

I would also add a third option to select the team and rights. In order to distinguish between read, write and admin access. So given the LDAP groups:
-  group1
-  group1-admin 
Then in gogs there would be 1 organization (group1) with 2 teams (admin and default?), the first one with admin rights by default. The organization and team are created after the configuration, and then the users are added to each team (at login time or via batch process)

**Comment by starwarsfan** [comment]
Usage of LDAP groups is a must have if using Gogs in a company. As @nestabur said: Automatic distinction between different permissions will be a great advantage.

**Comment by Unknwon** [comment]
Thanks @nestabur ! Will consider that definitely.

**Comment by mehrad-rafigh** [comment]
+1

**Comment by maxchill** [comment]
I'm interested in this as well.  However, I wanted to ask, does Gogs even have a self hosted Groups feature at the moment to build this LDAP group support on top of?  Perhaps that feature should be a seperate task from this?

**Comment by posix4e** [comment]
Same we will be moving to gitlab without ldap group integration. Basically we want some groups to see certain repos others cannot

**Comment by strk** [comment]
It would also be a useful feature for OSGeo, another organization using LDAP and evaluating a switch from trac to Gogs: http://www.osgeo.org

**Comment by ptman** [comment]
I would suggest this to work as a one-way sync. When an LDAP user logs in, a search is performed to find out the user's LDAP group memberships and the gogs groups memberships would be updated to match. How will groups be mapped to LDAP groups? Just by cn? or should groups/organizations have an LDAP-dn attribute? That could be nice if you want to keep a mix of ldap and gogs groups, but a lot of work with lots of groups.

**Comment by strk** [comment]
One-way is how email information also works, right ?
That is, if the mail field is updated in the LDAP server,
Gogs won't notice the update, correct ?

**Comment by ptman** [comment]
Do you mean one-time? One-way would be information moving only from LDAP -> gogs. Two-way would be LDAP -> gogs _and_ gogs -> LDAP, i.e. LDAP <-> gogs.

**Comment by strk** [comment]
Yes, I meant one-time. Is the email field only retrived one-time from LDAP to Gogs ?
I don't see a way to change email in Gogs so I guess it is also one-way.

**Comment by ptman** [comment]
It could check the information from LDAP on every login

**Comment by strk** [comment]
I've just checked and as of Gogs Version 0.9.13.0318 the local user email is not updated on a new login after changing the email in the LDAP server. I agree it would be recommendable to sync on login (new login, new sync). Then there'll need to be a way to enforce logging out every user as an administrator, when important changes are made to the LDAP directory.

**Comment by mbuchner** [comment]
+100 We also definitely need this feature - this would make "!the!" big difference against gitlab

**Comment by ptallada** [comment]
Yes, we would also consider moving from gitlab to gogs if we could sync LDAP->gog groups :)

**Comment by jortkoopmans** [comment]
I would love this feature!
LDAP group -> Gogs : one-way synchronization.
I'd think that it should check the groups on each login (when it does lookup in ldap anyway?).

Note there's generally 2 direct methods for LDAP/AD groups (that are usually both applied for consistency)

- Specify the 'MemberOf' or 'UniqueMemberOf' attributes on the User (DN). The attribute usually contains one or more CN's to groups (ex: CN=DevOps,OU=Development,DC=organisation,DC=com).

- Specify the 'Member' or 'UniqueMember' attributes on the Group (DN). The attribute usually contains  one or many more of either the full DN or UID of the User(s).

Depending on security settings for directly authenticated users and visibility of the groups, often an 'admin account' is used to be able to parse groups without restrictions.

Most of the times I've seen this requiring 3 parameters:
BindDN: full DN to authenticate the admin account
BindPassword: password to authenticate admin account
GroupSearchBase: DN from where to parse the group structure (ex: OU=Development,DC=organisation,DC=com)

Sometimes the attribute for members can also be specified (Member/UniqueMember).

**Comment by Ritbit** [comment]
For me this is absolutely key to have an oneway group-sync between LDAP and GOGS organisations.
What jortkoopmans describes above would work for me too. +100 if this gets completed.

**Comment by dobegor** [comment]
Also +1 to what @jortkoopmans said. We do want to switch to Gogs instead of Gitlab but there's no group from LDAP support in gogs.

**Comment by rgtx** [comment]
@jortkoopmans - I'd also add that a configurable group filter is necessary, for OUs that aren't necessarily clean (ie: non-group objects are - unfortunately - lumped into the same group search base).

'(objectClass=group)' (or anything else, as necessary)

GitLab EE also had - and fixed - a similar group filtering issue (GitLab support ticket 16230), for their LDAP group sync feature.

**Comment by skorhone** [comment]
In my opinion, it does not make sense to synchronize ldap groups to gogs. While we can live without this feature, here's a suggestion how to implement this.
 - Add new configuration items for ldap: membership filter (list of groups for given user) and user member filter (list of users for given group) Ignore support of nested groups for now
 - Add new configuration item for team: security groups. For ldap you would store full name of security group (listed/verified using member filter)
 - When listing team members, include such members from ldap that are not in any other team (again, use member filter)
 - When checking for user rights, prefer highest. User that belongs to administrative privileges should be granted them (even if same user is part of a team with read only privileges) (this is where you would prefer use of membership filter for performance reasons)
 - Finally when you have everything in place, implement caching (While ldap can ben fast - really fast, sizes of groups can be huge. Querying list of group members for each page view, might not make sense)

**Comment by jortkoopmans** [comment]
I'm seeing a new feature in the changelog that might be related, generated from pull req #4398?

**Comment by Unknwon** [comment]
@jortkoopmans right... but I think #4398 is a start, not yet complete all the requests in this thread maybe... since this is very long, not really sure :sweat:

**Comment by skorhone** [comment]
I can have a look at it next week. Having authentication and authorization code in one place makes it a bit cumbersome to follow. If I got it right, team name is used for searching ldap when checking user's permissions for repo.

**Comment by skorhone** [comment]
I see, the patch #4398 is not what I expected it to be. It allows controlling access to **gogs** using ldap groups. However, it does not allow settings read/write access of repositories using ldap groups.

I'm not quite sure, if this is the right way to go. These queries can't be reused later on for repository access / control. You will have to almost identical queries / filters defined in config. I'd rather see _group filter_ defined using the same parameter format that was used with _user filter_. This filter should be used with a _list of allowed groups_ and the query should be executed once per item - until allowed group is found. This would also allow one to implement restrictions based on groups without defining additional queries. And it would also allow one to reuse queries when implementing access control for repositories

**Comment by skorhone** [comment]
Phew, that was a lot of work. (Learning basic of new language in a day and fixing ldap-configuration to something that most people are more used to). I won't be bothering to post a pull request before I get repository to use ldap as an authorization source, but you can check/track changes here on my fork at ldap_repositories -branch.

There's bunch of new mandatory configuration items, as you can see from the picture:
![image](https://user-images.githubusercontent.com/4815560/27738575-5dc324c4-5db4-11e7-9ce2-d7f3e1ed9db1.png)

Some of them will get default values later on (most organizations use standard values)

While I am at it, I will propably look, if it would be possible not to copy user data to gogs, when using ldap for authentication and authorization (or at least sync data on every login). While I can't see any benefit for doing so, I can find many issues that this causes. I think gogs's internal database should be just one of many authentication / authorization sources. As in real world, also in gogs, accounts should be accessible using various authentication methods.

**Comment by skorhone** [comment]
I have now integrated ldap for access control and verified mechanism with manual testing (there's no unit tests in go / in this project?). Implemented algorithm is quite naive, but it's best I could come up with current domain model.

Important part is in models/access.go:

       // If this repository is linked to teams
        err := repo.GetOwner()
        if err != nil {
                return mode, err
        }
        if repo.Owner.IsOrganization() {
                // We need user for name. Id is just pointer to database
                user, err := GetUserByID(userID)
                if err != nil {
                        return mode, err
                }
                // We need to check, which authorization source user is bound to
                // It would be nice, if we had an interface to work with :-(
                // Otoh. LDAP is currently only true authorization source
                if user.LoginType == LOGIN_LDAP {
                        lsrc, err := GetLoginSourceByID(user.LoginSource)
                        if err != nil {
                                return mode, err
                        }
                        ldap := lsrc.LDAP()
                        // Get UID
                        uid, ok := ldap.GetUID(user.Name)
                        if !ok {
                                return mode, fmt.Errorf("could not get uid for: %v", user.Name)
                        }

                        // Reusing existing queries...
                        // We could use a query, that gets all teams for given repo
                        // But... I don't like the current domain model - so I won't do it :-)
                        teams, err := GetTeamsByOrgID(repo.Owner.ID)
                        if err != nil {
                                return mode, err
                        }
                        for _, team := range teams {
                                // Get access level from team, 
                                // if team grants higher access level, AND
                                // if team is actually linked to repository, AND
                                // if user is member of team
                                if team.Authorize > mode && HasTeamRepo(repo.Owner.ID, team.ID, repo.ID) && **ldap.IsMemberOf(uid, team.Name)** {
                                        mode = team.Authorize
                                }
                        }
                }
        }

From here you can see, that I chose to do 1:1 binding between team and group names. You can also see, that I did not bother to do any other modifications to code. I hinted in my #4599 that, full ldap integration with current domain model is just not worth the effort. It would take days to implement all required queries. And there's a high risk, that I would get something fundamentally wrong.

**Comment by skorhone** [comment]
It seems I would need to modify SearchRepositoryByName operation in order to get explore working properly with ldap authorization. Currently it's not utilizing access.go and it relies heavily on SQL operations (access table is joined against repository table) Because the search is returning a count, only reasonable thing to do is to fetch all repositories and filter them by access. If done correctly, this should work quite well, when gogs is managing reasonable number of repositories (N < 10000)

**Comment by skorhone** [comment]
This is almost done. I decided to move access control related code to access.go (there was some in repo.go). Now there's only one query left that needs to be changed in order to avoid N+1 problem. I just hope, that my decision to use IN queries (for avoiding N+1 query problem) will not cause problems on sites, that have very high number (thousands) of repos

**Comment by skorhone** [comment]
Implementation is now complete. For readability I actually ended up moving all repository related access level checks in repository.go and I modified interfaces for repository. Repository has now a new function called HasAccess with parameters user id & required access mode. For single repository code is quite clean, but for listing, it's a bit more complex (I opted for sql with nested queries instead of bombarding database with possibly thousand of queries or doing insanely long IN conditions)

**Comment by skorhone** [comment]
Final version of settings:

![image](https://user-images.githubusercontent.com/4815560/27907608-0633b952-6251-11e7-8c9d-6f37cd8f5e07.png)

As promised, there's bunch of defaults. Filters no longer have magical %v's - instead queries are built on runtime

**Comment by Unknwon** [comment]
@skorhone thanks for all the effort!

I think the OP means the LDAP group maps to an **organization not a team** in Gogs?

**Comment by skorhone** [comment]
While it is possible, I don't think it would work for most/everyone. It would solve single use case where everyone in given organization has full access to organization's repos. Having privileges bound to teams (ldap groups) provides much more flexibility (you can give everyone full access, or you can give read access to some and full access to others) btw. If you are interested in the update, I can try to merge changes done since I wrote it. I took liberty to do modifications in access checks (obsoleted access table in db)

**Comment by Unknwon** [comment]
@skorhone I'm not sure why "It would solve single use case where everyone in given organization has full access to organization's repos"?

**Comment by Ekleog** [comment]
Any news? it looks great! and indeed for our use case (association) we'd most likely rather have groups mapped to teams than to organizations.

I guess something like one LDAP <-> one organization and one group in this LDAP <-> one team would make most sense (with “one LDAP” potentially including metadata such as “server:port with group_filter”), but even then the first step of a single LDAP and mapping each group in it to a team would already be a great first step, that can later on be extended to support multiple LDAPs.

**Comment by maxguru** [comment]
Any updates on this?

**Comment by alexattws** [comment]
We've implemented mapping LDAP groups to teams in (all) organizations (by group/team name) for Gitea and published it here, if anyone is interested,
https://github.com/tws-inc/gitea-group-sync

**Comment by unknwon** [comment]
@alexattws can it be used for Gogs?

**Comment by alexattws** [comment]
If the Gogs API calls that are being used match Gitea API calls, then the answer is yes.

**Comment by unknwon** [comment]
@alexattws I have no idea (what APIs this tool uses), if it's not working with Gogs, posting here would be considered as unrelated.

**Comment by alexattws** [comment]
The code is here: https://github.com/tws-inc/gitea-group-sync/blob/master/gitea-group-sync.go

It uses the following Gitea calls,
/api/v1/users/search
/api/v1/admin/orgs
/api/v1/orgs/orgname/teams
/api/v1/teams/teamid/members

It is related to the discussion here, as far as I understand, as Gitea is a fork of Gogs and is somewhat compatible.  I haven't tested it myself with Gogs, but if it doesn't work, someone would potentially be able to make it work by making small changes.

**Comment by unknwon** [comment]
@alexattws Thanks! Unfortunately, I think Gogs does not have this endpoint: `/api/v1/teams/teamid/members`, but should be easy to add.
